R/unite.R

```diff
@@ -0,0 +1,61 @@
+#' Unite multiple columns into one by pasting strings together.
+#'
+#' @description
+#' This is a method for the tidyr `unite()` generic.
+#'
+#' @inheritParams tidyr::unite
+#' @examples
+#' library(tidyr)
+#'
+#' df <- lazy_dt(expand_grid(x = c("a", NA), y = c("b", NA)))
+#' df
+#'
+#' df %>% unite("z", x:y, remove = FALSE)
+#'
+#' # Separate is almost the complement of unite
+#' df %>%
+#'   unite("xy", x:y) %>%
+#'   separate(xy, c("x", "y"))
+#' # (but note `x` and `y` contain now "NA" not NA)
+# exported onLoad
+unite.dtplyr_step <- function(data, col, ..., sep = "_", remove = TRUE, na.rm = FALSE) {
+  if (is_true(na.rm)) {
+    abort("`na.rm` is not implemented in dtplyr")
+  }
+
+  .col <- as_name(enquo(col))
+
+  dots <- enquos(...)
+  if (length(dots) == 0) {
+    .cols <- data$vars
+  } else {
+    sim_data <- simulate_vars(data)
+    locs <- tidyselect::eval_select(expr(c(!!!dots)), sim_data, allow_rename = FALSE)
+    .cols <- data$vars[locs]
+  }
+
+  out <- mutate(ungroup(data), !!.col := paste(!!!syms(.cols), sep = sep), .before = !!sym(.cols[[1]]))
```

Comment:
Does not work correctly if cols are specified in a different order

```r
df %>% collect() %>% unite("z", c(y, x), remove = FALSE) %>% colnames()
#> "z" "x" "y"

df %>% unite("z", c(y, x), remove = FALSE) %>% .$vars
#> "x" "z" "y"
```

This should use something like `min(locs)`.

You could also produce a shorter query by using `relocate()` explicitly

```r
out <- mutate(ungroup(data), !!.col := paste(!!!syms(.cols), sep = sep))

  remove <- is_true(remove)
  if (remove) {
    .drop_cols <- setdiff(.cols, .col)
    out <- select(out, -tidyselect::all_of(.drop_cols))
  }
  out <- relocate(out, !!.col, .before = min(locs))
```